src/Microsoft.ML.TorchSharp/NasBert/TextClassificationTrainer.cs

```diff
@@ -0,0 +1,857 @@
+        {
+            // Verify that all required input columns are present, and are of the same type.
+            if (!inputSchema.TryFindColumn(_options.Sentence1ColumnName, out var sentenceCol))
+                throw _host.ExceptSchemaMismatch(nameof(inputSchema), "sentence", _options.Sentence1ColumnName);
+            if (sentenceCol.ItemType != TextDataViewType.Instance)
+                throw _host.ExceptSchemaMismatch(nameof(inputSchema), "sentence", _options.Sentence1ColumnName,
+                    TextDataViewType.Instance.ToString(), sentenceCol.GetTypeString());
+
+            if (!inputSchema.TryFindColumn(_options.LabelColumnName, out var labelCol))
+                throw _host.ExceptSchemaMismatch(nameof(inputSchema), "label", _options.LabelColumnName);
+            if (labelCol.ItemType != NumberDataViewType.UInt32)
+                throw _host.ExceptSchemaMismatch(nameof(inputSchema), "label", _options.LabelColumnName,
+                    NumberDataViewType.UInt32.ToString(), labelCol.GetTypeString());
+
+            if (_options.Sentence2ColumnName != default)
+            {
+                if (!inputSchema.TryFindColumn(_options.Sentence2ColumnName, out var sentenceCol2))
+                    throw _host.ExceptSchemaMismatch(nameof(inputSchema), "sentence2", _options.Sentence2ColumnName);
+                if (sentenceCol2.ItemType != TextDataViewType.Instance)
+                    throw _host.ExceptSchemaMismatch(nameof(inputSchema), "sentence2", _options.Sentence2ColumnName,
+                        TextDataViewType.Instance.ToString(), sentenceCol2.GetTypeString());
+            }
+        }
+    }
+
+    public sealed class SentenceClassificationTransformer : RowToRowTransformerBase
+    {
+        internal const string LoadName = "NASBERTTrainer";
+        internal const string UserName = "NASBERT Trainer";
+        internal const string ShortName = "NASBERT";
+        internal const string Summary = "NLP with NAS-BERT";
+
+        private readonly Device _device;
+        private readonly SentenceClassificationModel _model;
+        private readonly Vocabulary _vocabulary;
+        private readonly TextClassificationTrainer.Options _options;
+
+        private readonly string _predictedLabelColumnName;
+
+        public readonly SchemaShape.Column SentenceColumn;
+        public readonly SchemaShape.Column SentenceColumn2;
+        public readonly SchemaShape.Column LabelColumn;
+
+        internal const string LoaderSignature = "NASBERT";
+
+        internal SentenceClassificationTransformer(IHostEnvironment env, TextClassificationTrainer.Options options, SentenceClassificationModel model, Vocabulary vocabulary)
+           : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(SentenceClassificationTransformer)))
+        {
+            Contracts.Assert(((IHostEnvironmentInternal)env).FallbackToCpu != false || _device != CPU, "Fallback to CPU is false but no GPU detected");
```

Comment:
>((IHostEnvironmentInternal)env).FallbackToCpu != false || _device != CPU [](http://example.com/codeflow?start=29&length=72)

I am not sure I understand this assert. if the user decided to explicitly use _device = CPU, why we need to require FallbackToCpu has to be true?

Reply:
Fixed the _device initialization issue.

I think I also fixed the assert. It now checks if device is CPU and short circuits.